draft-ietf-quic-transport.md

```diff
@@ -2887,6 +3033,95 @@ by a client in protected packets, because it is certain that the server is able
 to decipher the packet.
 
 
+## ACK_ECN Frame {#frame-ack-ecn}
+
+A QUIC connection MUST keep counters for each ECN codepoint, recording
+the number of packets that were received with the corresponding ECN
+codepoint in the IP header. If the header is not readable from the
+application, the codepoint 00 (Not-ECT) MUST be assumed. If any packet
+are duplicated by the network then only the value of the ECN field of
+the packet copy first received SHALL be included in the counters, the ECN
+field value for a duplicate SHALL be ignored. This to prevent the
```

Comment:
Clarification question: considering the use of "SHALL," does this imply that an endpoint MUST detect all duplicates?

I understand that it is possible to implement such a scheme; an endpoint will keep track of the cut-off PN (the biggest PN that it would ignore) as well as all the PNs that it have seen above the cut-off PN. But I wonder if the intent here is to require all implementations to have such defense.

Reply:
Yes, that is the intention, to avoid getting ECN field values for duplicates added to the counters for packets that are included in any ACK. So if a receiver has included it in an ACK, it need to remember that this PN already have an ECN field valued added to the counters. But it doesn't need to detect duplicates for packet that are discarded directly after decryption, i.e. older than the cut-off PN.   

This is straightforward and fulfills the requirements that appears to currently exist. If #1405 concludes it would be easier to determine if there are other approaches including other ECN field value encodings as discussed in #1439.

Reply:
Thank you for the clarification. My two cents go to that this is an unnecessary burden, because IMO while idempotency is required for frame payloads, it is not required for ECN signal that is injected by an on-path device (even though we need some defense against injection attacks).

Reply:
This is incorrect. The requirements that currently exists in tsv is that ALL CEs must be counted because the network does not know if a packet is a duplicate or not!

Reply:
I believe the reason TCP counts all CEs is because TCP retransmits packets, not because of an effort to count packets the network inadvertently duplicated.  QUIC does not duplicate packets and instead uses new packet numbes for retransmissions, so I believe this embodies the intent of the requirement.

Reply:
No, that is not true. The network marks a packet as congested because it is congested and want to signal this to the endpoint. The network does not distinguish between retransmission or duplicates or anything because all these packets need the same amount of network resources and therefore cause congestion. RFC8085 even says 
"A receiver that enables the use of ECN for a UDP port MUST check
      the ECN field at the receiver for each UDP datagram that it
      receives on this port." 
and 
"This MUST report the
      presence of datagrams received with a CE-mark by providing a
      mechanism to feed this congestion information back to the sending
      application."
You cannot just ignore an CE because it's on a retransmission or a duplicated packet!

The problem with duplicates it that a packet could have been duplicated before or after it was CE marked. If it gets duplicated after it was marked, this is not a real congestion indication and if counted overstates the congestion. However, if it gets marked after it was duplicated, it is an indication of congestion. As you don't know when it got marked, you need to be conservative and account for it.

Reply:
Also QUIC does process duplicate packets to the point where it can detect that it is duplicated which probably already gives you some confidence that it is a valid quic packet and therefore you can also check the ECN IP mark at this point of time.

Reply:
Fair enough, then I think that text wasn't written with QUIC in mind.

As I mentioned previously, I think this is an edge case that is not a practical problem in real networks.  Unnecessary packet duplication is the real problem, not it's implications on ECN.

Reply:
I agree that this is an edge case and should hopefully not be a problem but saying in the spec that duplicated shall be ignored is not the right thing to do.

Reply:
Let's back up just a bit.

First, QUIC receivers are far less likely to receive duplicates because as @ianswett pointed out, retransmissions are not duplicates. As @mirjak notes, this doesn't mean that the network won't do stupid things, but that really is the exception, not the rule. Duplicates in QUIC are necessarily network duplication. This is a significant difference in terms of *how much* this problem matters. I think suppression is fine, since the tradeoff is between enabling an on-the-side performance attack vs. an infinitesimally more aggressive response to a CE marked packet.

Second, I agree with @kazuho that requiring duplicate detection should not be required, if only because I can cook up conditions under which an endpoint won't be able to do this without infinite state.  Consider an endpoint that nacks a packet (sends an ack with this packet in a gap), receives an ack for this nack, and moves its "left edge" past that PN. Now, it receives a packet with that PN. It cannot verify that this packet is a duplicate, since it may never have received the original.

I'd like to point out that this is ultimately a performance attack, and as with any other such attacks that use unauthenticated bits (ICMP?), we can provide some guidance to the receiver of a duplicate. I think the text here can changed a bit, with little to no real difference in impact.

I'll propose the following text: "The ECN counters that an endpoint reports include values of the ECN field only from packets that are subsequently newly acknowledged.  Note that an endpoint may newly acknowledge a duplicate packet if it no longer has state about the first receipt of the same packet. An endpoint may decrypt and drop a duplicate packet if it has state about the first receipt of the same packet, and in this case, the endpoint will not increase the ECN counters based on the duplicate packet."

Reply:
The problem with that text is that it assumes something quite dangerous: that a receiver might accept a packet when it isn't sure that it has received it before.  And content might be different.

What @ianswett suggested before is better: if you move the trailing edge forward, then packets that you believe to be behind that edge are dropped unconditionally.  Or, more concisely stated: you only accept packets that you know you haven't accepted before.  That errs on the side of dropping rather than accepting, which I believe to be safe.

Reply:
> The problem with that text is that it assumes something quite dangerous: that a receiver might accept a packet when it isn't sure that it has received it before. And content might be different.

I am not sure if the content could be different. I'd assume that the only case in which it could be different is when the sender reuses a packet number. I do not think that we need to detect that (and if we do, it should be a PROTOCOL_VIOLATION rather than a drop).

Reply:
@martinthomson it was not my intention to require infinite state. I will see if I can reformulate to say that for any packet within the window of PNs that you will accept to process at all you need to maintain duplication detection before adding the ECN field values to the counters. 

@mirjak TSVWG has asked if there was anything from the QUIC ECN work that was of general nature and should be discussed in TSVWG and I think this is one of those things. However, my opinion is that due to ECN's unauthenticated nature, it is reasonable to suppress network duplicates. Otherwise use of ECN makes one more open to attacks to drive down the congestion window for a connection than when not using it. Yes, the connection will be slightly less responsive to the congestion signal, how much is depending on the amount of network duplication that occur.